File: DataStructures/src/main/java/com/zhangdd/queue/CircleArrayQueue.java
package com.zhangdd.queue;

/**
 * //注意这里实现的环形队列，是空了一个空间，所以创建的大小为4时实际可以存放的数据个数是3个
 *
 * @author zhangdd on 2021/9/25
 */
public class CircleArrayQueue {

    private int front = 0;
    private int rear = 0;
    private final int[] arr;
    private final int maxSize;

    public CircleArrayQueue(int maxSize) {
        this.maxSize = maxSize;
        arr = new int[maxSize];
    }

    private boolean isFull() {
        return (rear + 1) % maxSize == front;
    }

    private boolean isEmpty() {
        return rear == front;
    }


    public void addQueue(int n) {
        if (isFull()) {
            System.out.println("当前队列已满，不能添加输入");
            return;
        }
        arr[rear] = n;
        rear = (rear + 1) % maxSize;
    }

    public int getQueue() {
        if (isEmpty()) {
            throw new RuntimeException("当前队列空，没有数据");
        }
        int data = arr[front];
        front = (front + 1) % maxSize;
        return data;
    }

    public int headQueue() {
        if (isEmpty()) {
            throw new RuntimeException("当前队列空，没有数据");
        }
        return arr[front];
    }

    public void showQueue() {
        if (isEmpty()) {
            System.out.println("当前队列空，没有数据");
            return;
        }
        int valueNumber = (rear - front + maxSize) % maxSize;
        for (int i = front, index = 0; index < valueNumber; i = (i + 1) % maxSize, index++) {
            System.out.printf("arr[%d]=%d\t", i, arr[i]);
        }
        System.out.println();
    }
}

File: python/DataStructures/sparsearry/CircleArrayQueue.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/1/7

环形队列，注意当前实现的队列可以存放的有效值个数为队列的长度-1
"""


class CircleArrayQueue:
    def __init__(self, max_size):
        self.max_size = max_size
        self.front = 0
        self.rear = 0
        self.queue = [None] * max_size

    def add_num(self, num):
        if self.is_ful():
            print("当前队列已满")
            return
        self.queue[self.rear] = num
        self.rear = (self.rear + 1) % self.max_size

    def get_num(self):
        if self.is_empty():
            print('当前队列为空')
            return
        print(self.queue[self.front])
        self.queue[self.front] = None
        self.front = (self.front + 1) % self.max_size

    def show(self):
        print(self.queue)

    def is_ful(self):
        return (self.rear + 1) % self.max_size == self.front

    def is_empty(self):
        return self.front == self.rear


queue = CircleArrayQueue(3)
queue.add_num(1)

queue.add_num(2)

queue.add_num(3)
queue.show()
command = ''
while True:
    print('a:add num')
    print('g:get num')
    print('s:show list')
    print('e:exit program')
    command = input('input command')
    if command == 'a':
        num = input('please input num')
        queue.add_num(num)
    elif command == 'g':
        queue.get_num()
    elif command == 's':
        queue.show()
    elif command == 'e':
        break
